pkg/reconciler/namespace/scheduler_test.go

```diff
@@ -0,0 +1,256 @@
+	return newClusterFixture(testLclusterName, otherTestClusterName)
+}
+
+func (f *clusterFixture) withReady() *clusterFixture {
+	conditions.MarkTrue(f.cluster, clusterv1alpha1.ClusterReadyCondition)
+	return f
+}
+
+func (f *clusterFixture) withUnscheduable() *clusterFixture {
+	f.cluster.Spec.Unschedulable = true
+	return f
+}
+
+func (f *clusterFixture) withFutureEvictionTime() *clusterFixture {
+	futureTime := metav1.NewTime(time.Now().Add(1 * time.Hour))
+	f.cluster.Spec.EvictAfter = &futureTime
+	return f
+}
+
+func (f *clusterFixture) withPassedEvictionTime() *clusterFixture {
+	pastTime := metav1.NewTime(time.Now().Add(-1 * time.Hour))
+	f.cluster.Spec.EvictAfter = &pastTime
+	return f
+}
+
+func newTestScheduler(clusters []*clusterv1alpha1.Cluster) namespaceScheduler {
+	return namespaceScheduler{
+		getCluster: func(name string) (*clusterv1alpha1.Cluster, error) {
+			for _, cluster := range clusters {
+				if clustertools.ToClusterAwareKey(cluster.ClusterName, cluster.Name) == name {
+					return cluster, nil
+				}
+			}
+			return nil, apierrors.NewNotFound(clusterv1alpha1.Resource("cluster"), name)
+		},
+		listClusters: func(selector labels.Selector) ([]*clusterv1alpha1.Cluster, error) {
+			return clusters, nil
+		},
+	}
+}
+
+func TestAssignCluster(t *testing.T) {
+	unknownClusterName := "unknown-cluster"
+
+	testCases := map[string]struct {
+		assignedCluster    string
+		schedulingDisabled bool
+		expectedCluster    string
+	}{
+		"scheduling disabled -> no change even for unknown cluster name": {
+			assignedCluster:    unknownClusterName,
+			schedulingDisabled: true,
+			expectedCluster:    unknownClusterName,
+		},
+		"valid assignment -> no change": {
+			assignedCluster: testClusterName,
+			expectedCluster: testClusterName,
+		},
+		"invalid assignment -> new assignment": {
+			assignedCluster: unknownClusterName,
+			expectedCluster: testClusterName,
+		},
+		"no assignment -> new assignment": {
+			expectedCluster: testClusterName,
+		},
+	}
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			clusters := []*clusterv1alpha1.Cluster{
+				defaultClusterFixture().withReady().cluster,
+			}
+			scheduler := newTestScheduler(clusters)
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					ClusterName: testLclusterName,
+					Labels: map[string]string{
+						clusterLabel: testCase.assignedCluster,
+					},
+				},
+			}
+			if testCase.schedulingDisabled {
+				ns.Labels[ScheduleDisabledLabel] = "true"
+			}
+			clusterName, err := scheduler.AssignCluster(ns)
+			require.NoError(t, err)
+			require.Equal(t, testCase.expectedCluster, clusterName)
+		})
+	}
+}
+
+func TestIsValidCluster(t *testing.T) {
+	testCases := map[string]struct {
+		cluster *clusterFixture
+		isValid bool
+	}{
+		"missing -> false": {},
+		"unready -> false": {
+			cluster: defaultClusterFixture(),
+		},
+		"ready -> true": {
+			cluster: defaultClusterFixture().withReady(),
```

Comment:
(not blocking, style preference): It seems like quite a lot of boilerplate to manually write/review/maintain above just so you can use this builder pattern to make the test case slightly shorter.

I tend to prefer just being verbose and putting the whole `v1alpha1.Cluster` struct here. Even though it might be somewhat noisy, it means I only have to look in one place to see what `withReady` means, and I don't have to maintain a separate testing builder mini-framework. I've seen these things grow to include dozens of fields, and it's really toilsome to strip back out.

Reply:
Shorter isn't the goal here. It's making my intent clear. For simple self-contained unit tests like this one, I'd much rather have to maintain some simple helpers to make it crystal clear what the cluster represents vs requiring a maintainer to have to confirm that the cluster matches expectation for each test case. 

As far as the potential for the helpers to metastasize out of proportion to their value, I think we can cross to that bridge when we come to it. If we keep our tests clean and focused I don't think the scope creep you mention is likely. If we start testing too many things at once, then sure, but I think that's a factoring problem in and of itself.

Reply:
That said, that's just me. If consensus is a preference not to use this pattern in testing, happy to revert.

Reply:
I am a fan of helpers, as long as there's not really anything "hidden" in them (in other words, they tend to do zero to minimal logic and just set fields based on function parameters).

Reply:
Sounds good to me.

Reply:
@imjasonh Please continue to keep me honest! I think your instinct to keep it simple is a good one, and I am generally in that camp too. If/when I stray I'll appreciate hearing about it from you.